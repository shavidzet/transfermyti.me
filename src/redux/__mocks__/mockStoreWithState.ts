import { ActionCreator, Middleware } from "redux";
import configureStore, { MockStore } from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { default as stateFixture } from "../__fixtures__/state";

export { MockStoreCreator, MockStore } from "redux-mock-store";

// This is the same as ReduxAction, but the type was changed to `any`, otherwise
// we'd have to explicitly specify every action `type` name when using jest-in-case:
export interface ReduxTestAction<TPayload = {}> {
  type: any;
  payload?: TPayload;
  error?: boolean;
  meta?: any;
}

export type ReduxTestActionCreator<TPayload> = ActionCreator<
  ReduxTestAction<TPayload>
  >;

// This is done to simplify the imports in tests (rather than have to import
// state/fixtures from the state.ts file and mockStoreWithState() function from
// another:
export const state = stateFixture;

/**
 * Creates an instance of a Redux store for mocking in tests.
 * @param stateChanges Updated entries in state (doesn't have to be entire state
 *                     since state fixture is used.
 * @param additionalMiddleware Any other middlewares to pass to configureStore
 *                             function.
 */
export function mockStoreWithState<TState>(
  stateChanges?: TState,
  additionalMiddleware: Array<Middleware> = [],
): MockStore {
  const sagaMiddleware = createSagaMiddleware();
  const mockStore = configureStore([sagaMiddleware, ...additionalMiddleware]);
  const validState = { ...stateFixture, ...stateChanges };
  return mockStore(validState);
}
