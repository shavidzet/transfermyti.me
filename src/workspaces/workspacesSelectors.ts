import { createSelector, createStructuredSelector } from "reselect";
import { selectIdToLinkedId } from "~/redux/reduxUtils";
import {
  Mapping,
  ReduxState,
  WorkspaceModel,
  WorkspacesByIdModel,
} from "~/typeDefs";

export const areWorkspacesFetchingSelector = (state: ReduxState): boolean =>
  state.workspaces.isFetching;

export const activeWorkspaceIdSelector = (state: ReduxState): string =>
  state.workspaces.activeWorkspaceId;

export const sourceWorkspacesByIdSelector = createSelector(
  (state: ReduxState): WorkspacesByIdModel => state.workspaces.source,
  (workspacesById): WorkspacesByIdModel => workspacesById,
);

export const sourceWorkspacesSelector = createSelector(
  sourceWorkspacesByIdSelector,
  (workspacesById): WorkspaceModel[] => Object.values(workspacesById),
);

export const targetWorkspacesByIdSelector = createSelector(
  (state: ReduxState): WorkspacesByIdModel => state.workspaces.target,
  (workspacesById): WorkspacesByIdModel => workspacesById,
);

export const includedSourceWorkspacesSelector = createSelector(
  sourceWorkspacesSelector,
  (workspaces): WorkspaceModel[] =>
    workspaces.filter(workspace => workspace.isIncluded),
);

export const sourceIncludedWorkspacesCountSelector = createSelector(
  includedSourceWorkspacesSelector,
  (includedWorkspaces): number => includedWorkspaces.length,
);

export const sourceWorkspacesForTransferSelector = createSelector(
  sourceWorkspacesSelector,
  (sourceWorkspaces): WorkspaceModel[] =>
    sourceWorkspaces.filter(workspace => workspace.isIncluded),
);

export const workspaceIdToLinkedIdSelector = createSelector(
  sourceWorkspacesByIdSelector,
  (sourceWorkspacesById): Record<string, string> =>
    selectIdToLinkedId(sourceWorkspacesById),
);

const targetWorkspacesSelector = createSelector(
  targetWorkspacesByIdSelector,
  (workspacesById): WorkspaceModel[] => Object.values(workspacesById),
);

export const missingTargetWorkspacesSelector = createSelector(
  includedSourceWorkspacesSelector,
  (includedSourceWorkspaces): WorkspaceModel[] =>
    includedSourceWorkspaces.filter(({ linkedId }) => linkedId === null),
);

const limitIdsToIncluded = (workspaces: WorkspaceModel[]): string[] =>
  workspaces.reduce((acc, workspace) => {
    if (!workspace.isIncluded) {
      return acc;
    }
    return [...acc, workspace.id];
  }, [] as string[]);

export const includedSourceWorkspaceIdsSelector = createSelector(
  sourceWorkspacesSelector,
  limitIdsToIncluded,
);

const includedTargetWorkspaceIdsSelector = createSelector(
  targetWorkspacesSelector,
  limitIdsToIncluded,
);

export const includedWorkspaceIdsByMappingSelector = createStructuredSelector<
  ReduxState,
  Record<Mapping, string[]>
>({
  source: includedSourceWorkspaceIdsSelector,
  target: includedTargetWorkspaceIdsSelector,
});
