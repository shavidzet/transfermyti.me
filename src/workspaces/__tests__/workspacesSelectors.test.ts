import cases from "jest-in-case";
import { state } from "~/redux/__mocks__/mockStoreWithState";
import * as workspacesSelectors from "../workspacesSelectors";

const TEST_STATE = {
  ...state,
  workspaces: {
    source: {
      "1001": {
        id: "1001",
        name: "Test Workspace",
        userIds: ["6001"],
        isAdmin: true,
        workspaceId: "1001",
        entryCount: 20,
        linkedId: "clock-workspace-01",
        isIncluded: true,
        memberOf: "workspaces",
      },
      "1002": {
        id: "1002",
        name: "Test Workspace 2",
        userIds: ["6001"],
        isAdmin: true,
        workspaceId: "1002",
        entryCount: 20,
        linkedId: null,
        isIncluded: false,
        memberOf: "workspaces",
      },
    },
    target: {
      "clock-workspace-01": {
        id: "clock-workspace-01",
        name: "Test Workspace",
        userIds: [],
        isAdmin: true,
        workspaceId: "clock-workspace-01",
        entryCount: 0,
        linkedId: "1001",
        isIncluded: true,
        memberOf: "workspaces",
      },
    },
    isFetching: false,
  },
};

describe("within workspacesSelectors", () => {
  cases(
    "selectors that directly access state return the correct value",
    options => {
      const result = options.selector(TEST_STATE);

      expect(result).toEqual(options.expected);
    },
    [
      {
        name: "areWorkspacesFetchingSelector returns state.isFetching",
        selector: workspacesSelectors.areWorkspacesFetchingSelector,
        expected: TEST_STATE.workspaces.isFetching,
      },
      {
        name:
          "activeWorkspaceIdSelector returns values from state.activeWorkspaceId",
        selector: workspacesSelectors.activeWorkspaceIdSelector,
        expected: TEST_STATE.workspaces.activeWorkspaceId,
      },
      {
        name: "sourceWorkspacesByIdSelector returns state.source",
        selector: workspacesSelectors.sourceWorkspacesByIdSelector,
        expected: TEST_STATE.workspaces.source,
      },
      {
        name:
          "sourceWorkspacesSelector returns array of values from state.source",
        selector: workspacesSelectors.sourceWorkspacesSelector,
        expected: Object.values(TEST_STATE.workspaces.source),
      },
      {
        name: "targetWorkspacesByIdSelector returns state.target",
        selector: workspacesSelectors.targetWorkspacesByIdSelector,
        expected: TEST_STATE.workspaces.target,
      },
    ],
  );

  cases(
    "the selectors match their snapshots",
    options => {
      const result = options.selector(TEST_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceWorkspacesSelector",
        selector: workspacesSelectors.includedSourceWorkspacesSelector,
      },
      {
        name: "for the sourceWorkspacesForTransferSelector",
        selector: workspacesSelectors.sourceWorkspacesForTransferSelector,
      },
      {
        name: "for the workspaceIdToLinkedIdSelector",
        selector: workspacesSelectors.workspaceIdToLinkedIdSelector,
      },
      {
        name: "for the missingTargetWorkspacesSelector",
        selector: workspacesSelectors.missingTargetWorkspacesSelector,
      },
      {
        name: "for the includedSourceWorkspaceIdsSelector",
        selector: workspacesSelectors.includedSourceWorkspaceIdsSelector,
      },
      {
        name: "for the includedWorkspaceIdsByMappingSelector",
        selector: workspacesSelectors.includedWorkspaceIdsByMappingSelector,
      },
    ],
  );

  test("the sourceIncludedWorkspacesCountSelector returns the count of included source workspaces", () => {
    const result = workspacesSelectors.sourceIncludedWorkspacesCountSelector(
      TEST_STATE,
    );

    expect(result).toBe(1);
  });
});
