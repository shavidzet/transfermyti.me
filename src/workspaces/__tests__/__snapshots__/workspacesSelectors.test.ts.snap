// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within workspacesSelectors the selectors match their snapshots for the includedSourceWorkspaceIdsSelector 1`] = `
Array [
  "1001",
]
`;

exports[`within workspacesSelectors the selectors match their snapshots for the includedSourceWorkspacesSelector 1`] = `
Array [
  Object {
    "entryCount": 20,
    "id": "1001",
    "isAdmin": true,
    "isIncluded": true,
    "linkedId": "clock-workspace-01",
    "memberOf": "workspaces",
    "name": "Test Workspace",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within workspacesSelectors the selectors match their snapshots for the includedWorkspaceIdsByMappingSelector 1`] = `
Object {
  "source": Array [
    "1001",
  ],
  "target": Array [
    "clock-workspace-01",
  ],
}
`;

exports[`within workspacesSelectors the selectors match their snapshots for the missingTargetWorkspacesSelector 1`] = `Array []`;

exports[`within workspacesSelectors the selectors match their snapshots for the sourceWorkspacesForTransferSelector 1`] = `
Array [
  Object {
    "entryCount": 20,
    "id": "1001",
    "isAdmin": true,
    "isIncluded": true,
    "linkedId": "clock-workspace-01",
    "memberOf": "workspaces",
    "name": "Test Workspace",
    "userIds": Array [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within workspacesSelectors the selectors match their snapshots for the workspaceIdToLinkedIdSelector 1`] = `
Object {
  "1001": "clock-workspace-01",
  "clock-workspace-01": "1001",
}
`;
